### PR TITLE
Housekeeping for partitioned DataObjects (#211)

### PR DESCRIPTION
**Pull request by zzeekk**
### What changes are included in the pull request?
see #211

### CODE REVIEW

**Comment by pgruetter** [comment]
Tried to release spark-extensions 2.0.12 for you, but the build fails. Will check again later.

About the implementation:
1. Is there a reason why TickTockHiveTableDataObject does not have housekeepingMode ?
2. How would I go about compacting timestamp partitions. I tried 
`archivePartitionExpression = "if( elements('dt') < add_months(current_date, -12), map('dt', trunc(elements('dt'), \"YEAR\" )), elements)"`
but that's not working
`Undefined function: 'elements'. This function is neither a registered temporary function nor a permanent function registered in the database 'default'.;`
Not sure about the exact syntax, but how do you use "elements" ?

**Comment by zzeekk** [comment]
@pgruetter: Thanks for the Release of spark-extensions. I checked - it still wants to deploy over bintray, which is shutdown. We have to migrate to deploy directly to sonatype / maven central. If we do this we might loose the history of the old artifacts. Would you mind? This happend to me with spark-temporalquery where i have created a new repository. Here we might deploy to the existing smartdatalake repository, but i dont know what happens to existing artifacts in the first deployment...

Good questions 
1) I'm not sure Housekeeping makes sense for TickTockHiveTableDataObject, because it is normally not partitioned, and for now we have only HousekeepingModes working with partitioned. Do you see a use case for Housekeeping with TickTockHiveTableDataObject? Should we at it even just for completness?
2) You can use `elements['dt']` or `elements.dt` to reference map entries. For now you only specified archivePartitionExpression, which moves a partition into another partition. To compact that partition later, you would also need to configure compactPartitionExpression which should return true for partitions which should be compacted. Note that in case of yearly archive partitions you must until all partitions of the year are archived for doing the compaction, as it is done only once.
3) thanks, deleted :-)

**Comment by pgruetter** [comment]
> If we do this we might loose the history of the old artifacts. Would you mind? 

So it's possible that old versions will be removed from maven central and only the latest version remains? Not sure if this is a problem or not, but it's not very clean. The new deployment works directly in maven with a profile, right? Could we execute a manual deployment of all the old versions? Anyway, we need to switch in any case as the current configuration will never work again so let's try and re-upload older versions if needed.

> 1. I'm not sure Housekeeping makes sense for TickTockHiveTableDataObject, because it is normally not partitioned, and for now we have only HousekeepingModes working with partitioned. Do you see a use case for Housekeeping with TickTockHiveTableDataObject? Should we at it even just for completness?

Never thought about if it makes sense or not, but I see quite a few partitioned TickTockHiveTables in my configurations. So yes, even just for completeness, let's implement it too. 

> 2. You can use `elements['dt']` or `elements.dt` to reference map entries. For now you only specified archivePartitionExpression, which moves a partition into another partition. To compact that partition later, you would also need to configure compactPartitionExpression which should return true for partitions which should be compacted. Note that in case of yearly archive partitions you must until all partitions of the year are archived for doing the compaction, as it is done only once.

So is the scaladoc wrong then? In Scaladoc for HousekeepingMode it says
`archivePartitionExpression = "if( elements('run_id') < runId - 1000, map('run_id', elements('run_id') div 1000), elements)"`
Does that need to be
`archivePartitionExpression = "if( elements['run_id'] < runId - 1000, map('run_id', elements['run_id'] div 1000), elements)"`
?

Did I understand the rest correctly: I want to move old monthly partitions into a bigger yearly partition. 
I can not do it as I did, moving everything older than 6 months, because it is only executed once. So executed today, the January partition is moved into a year 2021 partition. Executed next month, February is NOT archived because 2021 is already "done" ?

Still, for this particular case, I think the compactPartitionExpression is not needed as this only compacts the file within a partition. I'm trying to combine multiple "months" partitions into yearly partitions. Correct ?

**Comment by zzeekk** [comment]
Ok, I'll try to change the deployment to sonatype for spark-extensions.
Good catch, scala doc is wrong... i'll fix it.
And yes, `archivePartitionExpression = "if( elements.dt < add_months(current_date, -12), map('dt', trunc(elements.dt, \"YEAR\" )), elements)` should move monthly partitions after 12 months into yearly partitions. Then you might want to compact that yearly partition to reduce number of files after 24 months with `compactPartitionExpression = "if( elements.dt < add_months(current_date, -12)`.

File: docs/Reference.md [hunk 1]
```diff
@@ -346,3 +346,26 @@ The following list describes specific behaviour of DataObjects:
 Recipes for data pipelines with schema evolution:
 * CopyAction supports schema evolution if all data is processed by overwriting the whole output DataObject. It needs an output DataObject which doesn't have a fixed schema, e.g. HiveTableDataObject.
 * HistorizeAction & DeduplicateAction supports schema evolution for incremental data. They consolidate the existing data & schema of the output DataObject with a potentially new schema of the input DataObject. Then they overwrite the whole output DataObject. They need a TransactionalSparkTableDataObject as output, which doesn't have a fixed schema, e.g. TickTockHiveTableDataObject.
+
+## Housekeeping
+SmartDataLakeBuilder supports housekeeping for DataObjects by specifying the HousekeepingMode mode.
```
Comment (pgruetter) [inline docs/Reference.md:351]:
```suggestion
SmartDataLakeBuilder supports housekeeping for DataObjects by specifying the HousekeepingMode.
```

File: sdl-core/src/main/scala/io/smartdatalake/util/misc/CompactionUtil.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+   * Compacting hadoop partitions is not supported out-of-the-box by hadoop, as files need to be read with the correct format and written again.
+   * The following steps are used to compact partitions with Spark:
+   * 1. Check if compaction is already in progress by looking for a special file "_SDL_COMPACTING" in data objects root hadoop path. If it exists
+   *    and is not older than 12h exit compaction with Exception. Otherwise create/update special file "_COMPACTION". If the file is older than 12h
+   *    the compaction process is assumed to be crashed.
+   * 2. As step 5 is not atomic (delete and move are two operations), we need to check for possibly incomplete compactions of previous crashed runs and fix them.
+   *    Incomplete compactions are marked with a special file "_SDL_MOVING" in the temporary path.
+   *    Incomplete compacted partitions must be moved from temporary path to hadoop path (see step 5)
+   *    and marked as compacted (see step 6).
+   * 3. Filter already compacted partitions from given partitions by looking for "_SDL_COMPACTED" file, see step ...
```
Comment (pgruetter) [inline sdl-core/src/main/scala/io/smartdatalake/util/misc/CompactionUtil.scala:40]:
```suggestion
   * 3. Filter already compacted partitions from given partitions by looking for "_SDL_COMPACTED" file, see step 5
```
?

File: sdl-core/src/main/scala/io/smartdatalake/util/misc/CompactionUtil.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+   * 4. Data from partitions to be compacted is rewritten into a temporary path under this data objects hadoop path.
+   * 5. Partitions to be compacted are deleted from the hadoop path and moved from the temporary path to the hadoop path. This should be done one-by-one to reduce risk of data loss.
+   *    To recover in case of unexpected abort between delete and move, a special file "_SDL_MOVING" is created in temporary path before deleting hadoop path.
+   *    After moving the temporary path, this file is deleted again. Mark compacted partitions by creating a special file "_SDL_COMPACTED" and
+   * 6  Delete "_SDL_COMPACTING" file created in step 1.
```
Comment (pgruetter) [inline sdl-core/src/main/scala/io/smartdatalake/util/misc/CompactionUtil.scala:45]:
```suggestion
   * 6.  Delete "_SDL_COMPACTING" file created in step 1.
```

File: sdl-core/src/main/scala/io/smartdatalake/util/misc/SparkExpressionUtil.scala [hunk 4]
```diff
@@ -102,6 +102,25 @@ private[smartdatalake] object SparkExpressionUtil {
     }
   }
 
+  /**
+   * Evaluate an expression against a each entry of a list of case class instances
```
Comment (pgruetter) [inline sdl-core/src/main/scala/io/smartdatalake/util/misc/SparkExpressionUtil.scala:106]:
```suggestion
   * Evaluate an expression against a each entry of a list of case class instances
```
```suggestion
   * Evaluate an expression against each entry of a list of case class instances
```

**Review by pgruetter — APPROVED** [review]
Just a few minor comments. Couldn't test the functionality yet, but fine for me to merge already.

**Pull request merged**